Spring 03 exercise

otussandra opened the pull request.

otussandra committed:
homework 3

homework 3

otussandra committed:
homework3

Перенести приложение для тестирования студентов на Spring Boot

Цель:
Цель: использовать возможности Spring Boot, чтобы разрабатывать современные приложения, так, как их сейчас и разрабатывают. Результат: Production-ready приложение на Spring Boot

Это домашнее задание выполняется на основе предыдущего.

Создать проект, используя Spring Boot Initializr (https://start.spring.io)
Перенести приложение проведения опросов из прошлого домашнего задания.
Перенести все свойства в application.yml
Локализовать выводимые сообщения и вопросы (в CSV-файле). MesageSource должен быть из автоконфигурации Spring Boot.
Сделать собственный баннер для приложения.
Перенести тесты и использовать spring-boot-test-starter для тестирования

otussandra committed:
homework3

не получается прочитать свойства из application.yml, подскажите, пожалуйста в чем ошибка

otussandra committed:
Homework3

нашла опечатку . вопрос снят.

otussandra committed:
Homework 3

Перенести приложение для тестирования студентов на Spring Boot
Создать проект, используя Spring Boot Initializr (https://start.spring.io)
Перенести приложение проведения опросов из прошлого домашнего задания.
Перенести все свойства в application.yml
Локализовать выводимые сообщения и вопросы (в CSV-файле). MesageSource должен быть из автоконфигурации Spring Boot.
Сделать собственный баннер для приложения.
Перенести тесты и использовать spring-boot-test-starter для тестирования
если Ваш язык отличается от русского и английского - локализовать в нём.

stvort commented on spring-03-exercise/src/main/java/ru/otus/spring/config/QuestionnaireConfig.java:
```diff
@@ -0,0 +1,59 @@
+        String language =System.getProperty("user.language");
+        locale = country+"_"+language;
+        return locale;
+    }
```
В таком случае для чего локаль это поле?

stvort commented on spring-03-exercise/src/main/java/ru/otus/spring/config/QuestionnaireConfig.java:
```diff
@@ -0,0 +1,59 @@
+    public String getQuestionnaireSource() {
+        String loc = this.getLocale();
+        String ss = questionnaireSource.replace("{locale}",loc);
+        return questionnaireSource.replace("{locale}",loc);
```
Зачем два раза делать `replace`?

stvort commented on spring-03-exercise/src/main/java/ru/otus/spring/dao/QuestionnaireDaoCSV.java:
```diff
@@ -0,0 +1,79 @@
+
+    public QuestionnaireDaoCSV(QuestionnaireConfig config) {
+        this.config = config;
+        this.questionnaireSource = config.getQuestionnaireSource();
```
Опять же, поле конфига не используется за пределами конструктора. Лучше как раз имя файла на лету получать

stvort commented on spring-03-exercise/src/main/java/ru/otus/spring/dao/QuestionnaireDaoCSV.java:
```diff
@@ -0,0 +1,79 @@
+public class QuestionnaireDaoCSV implements QuestionnaireDao {
+    private final String questionnaireSource;
+    private final char delimiter;
+    private final QuestionnaireConfig config;
```
Не уверен, что дао нужен прямо весь конфиг, возможно стоит ему дать только часть. Хотя бы с помощью реализации конфигом разных интерфейсов

otussandra replied:
не очень понимаю как это сделать. поделить конфиг на 2 файла?

stvort replied:
Можно и так. Я же писал про два интерфейса у одного конфига. Тогда в дао мы по одному будем внедрять, а в другие классы по другому

stvort commented on spring-03-exercise/src/main/java/ru/otus/spring/service/QuestionnaireServiceImpl.java:
```diff
@@ -0,0 +1,64 @@
+public class QuestionnaireServiceImpl implements QuestionnaireService {
+    private final QuestionnaireDao dao;
+    private final IOService ioService;
+    private final QuestionnaireConfig config;
```
Тут конфиг тоже не используется) Тоже лучше использовать именно его, а не сохранять в поля его составные части. Так сменой локали в конфиге можно будет сменить ее во всем приложении

stvort commented on spring-03-exercise/src/main/java/ru/otus/spring/api/IOService.java:
```diff
@@ -0,0 +1,11 @@
+import org.springframework.context.MessageSource;
+
+public interface IOService {
+    void out(String messagecode, MessageSource messagesource,String[] message);
```
Так мы выдаем детали реализации. Лучше оставить сервис ввода-вывода как он был, но сделать еще два. Первый будет работать с конфигом и локалью из конфига, а второй использовать первый + сервис ввода-вывода

stvort commented on spring-03-exercise/src/test/java/ru/otus/spring/QuestionnaireResultTest.java:
```diff
@@ -0,0 +1,22 @@
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+@SpringBootTest
+public class QuestionnaireResultTest {
```
`@SpringBootTest` поднимает контекст, который для данного теста не нужен

stvort replied:
>`@SpringBootTest` поднимает контекст, который для данного теста не нужен

stvort replied:
>`@SpringBootTest` поднимает контекст, который для данного теста не нужен

otussandra committed:
homework3

исправление замечаний

stvort commented on spring-03-exercise/src/main/java/ru/otus/spring/api/MessageService.java:
```diff
@@ -0,0 +1,21 @@
+import java.util.Locale;
+
+@Service
+public class MessageService {
```
Лучше с интерфейсом сделать

stvort commented on spring-03-exercise/src/main/java/ru/otus/spring/api/MessageService.java:
```diff
@@ -0,0 +1,21 @@
+    private final Locale locale;
+    private final MessageSource msg;
+    public MessageService(MessageSource msg,  QuestionnaireConfig config){
+        this.locale = Locale.forLanguageTag(config.getLocale());
+        this.msg    =   msg;
```
Локаль сразу из конфига можно отдавать в виде `Locale`. И лучше это сделать в момент вызова метода `getMessage`. Тогда сменив локаль в конфиге, мы ее на лету сменим во всем приложении

stvort commented on spring-03-exercise/src/main/java/ru/otus/spring/api/MessageService.java:
```diff
@@ -0,0 +1,21 @@
+        this.msg    =   msg;
+
+    }
+    public String getMessage(String msgCode,  String[] message){
```
Вряд ли второй аргумент это `message`, скорее `args`. И его лучше сделать вараргами, чтобы можно было вызывать так: `getMessage("any.message.code", arg1, arg2, arg3)`

stvort commented on spring-03-exercise/src/main/java/ru/otus/spring/Spring03ExerciseApplication.java:
```diff
@@ -0,0 +1,23 @@
+
+	public static void main(String[] args) {
+		ApplicationContext ctx =SpringApplication.run(Spring03ExerciseApplication.class, args);
+		MessageSource msg = ctx.getBean(MessageSource.class);
+		QuestionnaireService service = ctx.getBean(QuestionnaireService.class);
+		IOService ioService = ctx.getBean(IOService.class);
+		service.startQuestionnaire();
```
Нам тут только один бин нужно достать

stvort commented on spring-03-exercise/src/main/java/ru/otus/spring/service/QuestionnaireServiceImpl.java:
```diff
@@ -0,0 +1,61 @@
+
+@Service
+public class QuestionnaireServiceImpl implements QuestionnaireService {
+    private final QuestionnaireDao dao;
+    private final IOService ioService;
+    private final QuestionnaireConfig config;
+    private final QuestionnaireRespondentIntroduceService questrespintroduceserv;
```
Все же принято писать переменные/поля в lowerCamelCase. В данном случае я про то что слова в названии принято выделять первой большой буквой

stvort commented on spring-03-exercise/src/main/java/ru/otus/spring/service/QuestionnaireServiceImpl.java:
```diff
@@ -0,0 +1,61 @@
+                questionresultserv.showQuestionnaireResult(config.getPassingScore(), new QuestionnaireResult(respondent, rightAnswerCounter));
+
+            } catch (RespondentIntroduceException e) {
+                ioService.out(messageService.getMessage("strings.introduceError",new String[] {""}));
```
Ага) Теперь еще завернуть `ioService` + `messageService` в `LocalizedIOService` и вообще красота будет)

stvort commented on spring-03-exercise/src/main/resources/application.yml:
```diff
@@ -0,0 +1,8 @@
+  delimiter: ;
+  questionnaireSource: /Questionnaire_{locale}.csv
+  passingScore: 3
+  locale: ru
```
У меня, если написать "en" вместо ru не происходить переключения на английский т.к. у меня русская системная локаль

otussandra committed:
some changes for homework3

otussandra committed:
homework3

исправление для выбора локали, у меня вроде начало переключаться при смене в конфиге на en

otussandra committed:
Homework3

доработка сервиса локализованного ввода

stvort commented on spring-03-exercise/src/main/java/ru/otus/spring/api/MessageService.java:
```diff
@@ -0,0 +1,21 @@
+
+
+@Service
+public class MessageService {
```
>Лучше с интерфейсом сделать

stvort commented on spring-03-exercise/src/main/java/ru/otus/spring/api/LocalizedIOService.java:
```diff
@@ -0,0 +1,16 @@
+import org.springframework.stereotype.Service;
+
+@Service
+public class LocalizedIOService{
```
Это тоже)

stvort commented on spring-03-exercise/src/main/java/ru/otus/spring/service/ExecuteQuestionnaireService.java:
```diff
@@ -0,0 +1,69 @@
+@Service
+public class ExecuteQuestionnaireService {
+    private final IOService ioService;
+    private final LocalizedIOService localizedIOService;
```
Ну и т.к. `LocalizedIOService` это у нас фасад для `MessageService` + `IOService`, то можно чтобы он умел делать все что умеют делать оба сервиса (прокидывая до них вызовы) + локализованный вывод. Это сократит количество зависимостей в сервисах, куда сейчас приходят  `IOService` и `LocalizedIOService`

stvort commented on spring-03-exercise/src/main/java/ru/otus/spring/service/QuestionnaireRespondentIntroduceService.java:
```diff
@@ -0,0 +1,41 @@
+import java.util.InputMismatchException;
+
+@Service
+public class QuestionnaireRespondentIntroduceService {
```
Давайте всем сервисам сделаем интерфейсы

stvort commented on spring-03-exercise/src/main/java/ru/otus/spring/service/QuestionnaireServiceImpl.java:
```diff
@@ -0,0 +1,56 @@
+    public void startQuestionnaire() {
+        try {
+            Questionnaire questionnaire = dao.loadQuestionnaire();
+            int rightAnswerCounter;
```
Зачем объявлять переменную вне блока `try`?

stvort commented on spring-03-exercise/src/main/java/ru/otus/spring/service/QuestionnaireServiceImpl.java:
```diff
@@ -0,0 +1,56 @@
+                questionResultServ.showQuestionnaireResult(config.getPassingScore(), new QuestionnaireResult(respondent, rightAnswerCounter));
+
+            } catch (RespondentIntroduceException e) {
+                localizedIOService.printMessage("strings.introduceError","");
```
В варарги можно не передавать ничего, если не нужно. В этом одно из их преимуществ

stvort commented on spring-03-exercise/src/main/java/ru/otus/spring/service/QuestionnaireResultService.java:
```diff
@@ -0,0 +1,28 @@
+        this.localizedIOService = localizedIOService;
+    }
+
+    public void showQuestionnaireResult(int passingScore,QuestionnaireResult questionnaireResult) {
```
`passingScore` можно включит в `QuestionnaireResult ` или сразу внедрить в `QuestionnaireResultService`

stvort commented on spring-03-exercise/src/main/java/ru/otus/spring/config/QuestionnaireConfig.java:
```diff
@@ -0,0 +1,60 @@
+
+@ConfigurationProperties(prefix = "properties")
+@Component
+public class QuestionnaireConfig {
```
Сейчас, в `MessageService`, `QuestionnaireDaoCSV` и `QuestionnaireServiceImpl` приходит один и тот же класс конфигурации (`QuestionnaireConfig`). Получается они связаны друг с другом через конфиг. При этом каждому из них нужна только часть. Если сделать, чтобы `QuestionnaireConfig` реализовал три интерфейса и внедрялся в компоненты по ним, то мы эту связь разорвем и наши классы смогут быть переиспользованы в других местах, независимо друг от друга

otussandra committed:
Homework3

исправление замечаний
добавлены интерфейсы для конфига и сервисов

stvort commented on spring-03-exercise/src/main/java/ru/otus/spring/api/IOService.java:
```diff
@@ -0,0 +1,9 @@
+package ru.otus.spring.api;
+
+import org.springframework.context.MessageSource;
```
За неиспользованными импортами тоже желательно следить, если будем переносить интерфейс в библиотеку или еще куда, где не будет `MessageSource` будет ошибка компиляции

stvort commented on spring-03-exercise/src/main/java/ru/otus/spring/service/QuestionnaireRespondentIntroduceServiceImpl.java:
```diff
@@ -0,0 +1,39 @@
+            if (lastName.length() > 0 && firstName.length() > 0) {
+                return new Respondent(lastName, firstName);
+            } else {
+                throw new RespondentIntroduceException("You didn't introduce yourself." + "\n", null);
```
Выглядит как попытка общаться с пользователем через исключения)

stvort commented on spring-03-exercise/src/main/java/ru/otus/spring/service/QuestionnaireResultServiceImpl.java:
```diff
@@ -0,0 +1,30 @@
+@Service
+public class QuestionnaireResultServiceImpl implements QuestionnaireResultService{
+    private final LocalizedIOService localizedIOService;
+    private final QuestionnaireResultConfig config;
```
Это не используется)

stvort commented on spring-03-exercise/src/main/java/ru/otus/spring/service/QuestionnaireResultServiceImpl.java:
```diff
@@ -0,0 +1,30 @@
+public class QuestionnaireResultServiceImpl implements QuestionnaireResultService{
+    private final LocalizedIOService localizedIOService;
+    private final QuestionnaireResultConfig config;
+    private int passingScore;
```
`final`?
